File: lib/rag/embedder.ts
import OpenAI from "openai";

const MODEL = "text-embedding-3-small";
const BATCH_SIZE = 100; // OpenAI max per request

export type EmbeddingResult = {
  embeddings: number[][];
  tokenCount: number;
};

export type QueryEmbeddingResult = {
  embedding: number[];
  tokenCount: number;
};

/** Minimal interface for the OpenAI embeddings client (for DI/testing). */
export type EmbeddingClient = {
  embeddings: {
    create: (params: {
      model: string;
      input: string[];
    }) => Promise<{
      data: { embedding: number[]; index: number }[];
      usage: { prompt_tokens: number; total_tokens: number };
    }>;
  };
};

let defaultClient: EmbeddingClient | null = null;

/**
 * Get or create the default OpenAI client.
 * Override with setEmbeddingClient() for testing.
 */
export function getEmbeddingClient(): EmbeddingClient {
  if (!defaultClient) {
    defaultClient = new OpenAI() as unknown as EmbeddingClient;
  }
  return defaultClient;
}

/** Override the embedding client (useful for testing). */
export function setEmbeddingClient(client: EmbeddingClient | null): void {
  defaultClient = client;
}

/**
 * Generate embeddings for a batch of texts.
 * Automatically splits into sub-batches of 100 (OpenAI limit).
 * Returns embeddings in the same order as input texts.
 */
export async function embedTexts(texts: string[]): Promise<EmbeddingResult> {
  if (texts.length === 0) {
    return { embeddings: [], tokenCount: 0 };
  }

  const client = getEmbeddingClient();
  const allEmbeddings: number[][] = [];
  let totalTokens = 0;

  // Process in batches
  for (let i = 0; i < texts.length; i += BATCH_SIZE) {
    const batch = texts.slice(i, i + BATCH_SIZE);
    const response = await client.embeddings.create({
      model: MODEL,
      input: batch,
    });

    // Sort by index to preserve order
    const sorted = [...response.data].sort((a, b) => a.index - b.index);
    for (const item of sorted) {
      allEmbeddings.push(item.embedding);
    }

    totalTokens += response.usage.prompt_tokens;
  }

  return {
    embeddings: allEmbeddings,
    tokenCount: totalTokens,
  };
}

/**
 * Generate a single embedding for a search query.
 */
export async function embedQuery(
  query: string
): Promise<QueryEmbeddingResult> {
  const result = await embedTexts([query]);
  return {
    embedding: result.embeddings[0],
    tokenCount: result.tokenCount,
  };
}

File: services/ingestion/src/embedder.py
from dataclasses import dataclass, field

from src.config import settings

BATCH_SIZE = settings.embedding_batch_size
MODEL = settings.embedding_model


@dataclass
class EmbeddingResult:
    embeddings: list[list[float]] = field(default_factory=list)
    token_count: int = 0


_client = None


def get_embedding_client():
    global _client
    if _client is None:
        import openai

        _client = openai.OpenAI().embeddings
    return _client


def set_embedding_client(client) -> None:
    global _client
    _client = client


def embed_texts(texts: list[str]) -> EmbeddingResult:
    if not texts:
        return EmbeddingResult()

    client = get_embedding_client()
    all_embeddings: list[list[float]] = []
    total_tokens = 0

    for i in range(0, len(texts), BATCH_SIZE):
        batch = texts[i : i + BATCH_SIZE]
        response = client.create(model=MODEL, input=batch)

        sorted_data = sorted(response.data, key=lambda x: x.index)
        for item in sorted_data:
            all_embeddings.append(item.embedding)
        total_tokens += response.usage.prompt_tokens

    return EmbeddingResult(embeddings=all_embeddings, token_count=total_tokens)
